src/runtime_src/core/pcie/driver/linux/xocl/userpf/xocl_kds.c

```diff
@@ -349,6 +385,226 @@ static int xocl_context_ioctl(struct xocl_dev *xdev, void *data,
 	return ret;
 }
 
+int xocl_create_hw_context(struct xocl_dev *xdev, struct drm_file *filp,
+		struct drm_xocl_create_hw_ctx *hw_ctx_args, uint32_t slot_id)
+{
+	struct kds_client *client = filp->driver_priv;
+	struct kds_client_hw_ctx *hw_ctx = NULL;
+	uuid_t *xclbin_id = NULL;
+	int ret = 0;
+
+	if (!client)
+		return -EINVAL;
+
+	ret = XOCL_GET_XCLBIN_ID(xdev, xclbin_id);
+	if (ret)
+		return ret;
+
+	mutex_lock(&client->lock);
+
+	hw_ctx = kds_alloc_hw_ctx(client, xclbin_id, slot_id);
+	if (!hw_ctx) {
+		ret = -EINVAL;
+		goto error_out;
+	}
+
+	/* Lock the bitstream. Unlock the same in destroy context */
+	ret = xocl_icap_lock_bitstream(xdev, xclbin_id);
+	if (ret) {
+		kds_free_hw_ctx(client, hw_ctx);
+		ret = -EINVAL;
+		goto error_out;
+	}
+
+	hw_ctx_args->hw_context = hw_ctx->hw_ctx_idx;
+
+error_out:
+	mutex_unlock(&client->lock);
+	XOCL_PUT_XCLBIN_ID(xdev);
+	return ret;
+}
+
+int xocl_destroy_hw_context(struct xocl_dev *xdev, struct drm_file *filp,
+		struct drm_xocl_destroy_hw_ctx *hw_ctx_args)
+{
+        struct kds_client_hw_ctx *hw_ctx = NULL;
+	struct kds_client *client = filp->driver_priv;
+        int ret = 0;
+
+        mutex_lock(&client->lock);
+
+        hw_ctx = kds_get_hw_ctx_by_id(client, hw_ctx_args->hw_context);
+        if (!hw_ctx) {
+                userpf_err(xdev, "No valid HW context is open");
+		mutex_unlock(&client->lock);
+                return -EINVAL;
+        }
+
+	/* Unlock the bitstream for this HW context if no reference is there */
+	(void)xocl_icap_unlock_bitstream(xdev, hw_ctx->xclbin_id);
+
+	ret = kds_free_hw_ctx(client, hw_ctx);
+
+	mutex_unlock(&client->lock);
+
+        return ret;
+}
+
+static int
+xocl_cu_ctx_to_info(struct xocl_dev *xdev, struct drm_xocl_open_cu_ctx *cu_args,
+	struct kds_client_hw_ctx *hw_ctx, struct kds_client_cu_info *cu_info)
+{
+        uint32_t slot_hndl = hw_ctx->slot_idx;
+        struct kds_sched *kds = &XDEV(xdev)->kds;
+        char *kname_p = cu_args->cu_name;
+        struct xrt_cu *xcu = NULL;
+        char iname[64];
```

Comment:
Magic number 64.

Reply:
Why in future PRs.  
You could easily do one of the following:
1. Add a const static int variable right before this array defining what the value is.
2. Add a define statement indicating what the value is.
3. Add a comment on what is so magically about 64.

Reply:
Fixed. Thanks